imagery/i.svm.predict/main.c

```diff
@@ -0,0 +1,419 @@
+
+    if (G_unqualified_name
+        (opt_sigfile->answer, NULL, name_sigfile, mapset_sigfile) == 0)
+        strcpy(mapset_sigfile, G_mapset());
+    if (!I_find_signature2
+        (I_SIGFILE_TYPE_LIBSVM, name_sigfile, mapset_sigfile))
+        G_fatal_error(_("Signature file <%s@%s> not found"), name_sigfile,
+                      mapset_sigfile);
+
+    if (G_unqualified_name
+        (opt_values->answer, G_mapset(), name_values, mapset_values) < 0)
+        G_fatal_error(_("<%s> does not match the current mapset"),
+                      mapset_values);
+    if (G_legal_filename(name_values) < 0)
+        G_fatal_error(_("<%s> is an illegal file name"), name_values);
+
+    /* Get bands */
+    if (opt_subgroup->answer) {
+        if (!I_get_subgroup_ref2
+            (name_group, opt_subgroup->answer, mapset_group, &group_ref)) {
+            G_fatal_error(_("There was an error reading subgroup <%s> in group <%s@%s>"),
+                          opt_subgroup->answer, name_group, mapset_group);
+        }
+    }
+    else {
+        if (!I_get_group_ref2(name_group, mapset_group, &group_ref)) {
+            G_fatal_error(_("There was an error reading group <%s@%s>"),
+                          name_group, mapset_group);
+        }
+    }
+    if (group_ref.nfiles <= 0) {
+        if (opt_subgroup->answer)
+            G_fatal_error(_("Subgroup <%s> in group <%s@%s> contains no raster maps."),
+                          opt_subgroup->answer, name_group, mapset_group);
+        else
+            G_fatal_error(_("Group <%s@%s> contains no raster maps."),
+                          name_group, mapset_group);
+    }
+    semantic_labels_group = G_malloc(group_ref.nfiles * sizeof(char *));
+    for (int n = 0; n < group_ref.nfiles; n++) {
+        semantic_labels_group[n] =
+            Rast_get_semantic_label_or_name(group_ref.file[n].name,
+                                            group_ref.file[n].mapset);
+    }
+
+    I_get_signatures_dir(sigfile_dir, I_SIGFILE_TYPE_LIBSVM);
+    /* Reorder rasters to match the training order */
+    misc_file =
+        G_fopen_old_misc(sigfile_dir, "version", name_sigfile,
+                         mapset_sigfile);
+    if (fscanf(misc_file, "%d", &sigfile_version) != 1) {
+        G_fatal_error(_("Invalid signature file"));
+    }
+    fclose(misc_file);
+    /* Current version number is 1 */
+    if (sigfile_version != 1) {
+        G_fatal_error(_("Invalid signature file version"));
+    }
+
+    /* Reorder group items to match order from the signature file */
+    misc_file =
+        G_fopen_old_misc(sigfile_dir, "semantic_label", name_sigfile,
+                         mapset_sigfile);
+    if (!misc_file)
+        G_fatal_error(_("Unable to read signature file '%s'."), name_sigfile);
+    names_ordered = G_malloc(group_ref.nfiles * sizeof(char *));
+    mapsets_ordered = G_malloc(group_ref.nfiles * sizeof(char *));
+    while (fscanf(misc_file, "%" XSTR(GNAME_MAX) "s", semantic_label) == 1) {
```

Comment:
Potential overflow, resulting string from fscan is not counting with `\n` char.

```suggestion
    char frmt[10];
    snprintf(frmt, sizeof(frmt), "%%%ds", GNAME_MAX - 1);
    while (fscanf(misc_file, frmt, semantic_label) == 1) {
```

If you go with this suggestion, the `XSTR`/`STR` macros are no more needed.

Reply:
Thanks for the suggestion. Your version is commited.